lib/hexpm/accounts/web_auth.ex

```diff
@@ -0,0 +1,113 @@
+defmodule Hexpm.Accounts.WebAuth do
+  use Hexpm.Context
+
+  @moduledoc false
+
+  # A pool for storing and validating web auth requests.
+
+  alias Hexpm.Accounts.WebAuthRequest
+  alias Hexpm.Accounts.Keys
+
+  # `device_code` refers to the code assigned to a client to identify it
+  # `user_code` refers to the code the user enters to authorize a client
+  # `scope` refers to the permissions granted to a key
+  # `scopes` refers to the list of scopes that are allowed in a web auth request
+  # `key_permission` is the default permissions given to generate a hex api key
+  # `key_params` is the default params to generate a hex api key
+
+  @key_permission %{domain: "api", resource: nil}
+  @key_params %{name: nil, permissions: [@key_permission]}
+
+  @doc """
+  Adds a web auth request to the pool and returns the response.
+
+  ## Params
+
+    - `scope` - Scope of the key to be generated. One of read and write.
+  """
+
+  def get_code(scope) do
+    device_code = "foo"
+    user_code = "bar"
```

Comment:
We really ought to find a way of generating these values.

Any suggestions @wojtekmach and @ericmj ?

Reply:
user code is easier, it should be relatively easy to type so I'd go with what GitHub has, I believe 8 letters and/or digits? Device code must have higher entropy though. I'd also check how roughly GitHub does this, how long is the secret. How about this for device code?

```elixir
iex(19)> :crypto.strong_rand_bytes(16) |> Base.url_encode64(padding: false)
"5bbMz-sRFl_udEZTUC9s3g"
```

Reply:
> I believe 8 letters and/or digits?

Yes.

> I'd also check how roughly GitHub does this, how long is the secret.

They don't explain how they generate the code, but they do mention that it is 40 chars long.

Looking at it, I think it is SHA, but I'm not very knowledgeable in cryptography.

See [the docs](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#step-1-app-requests-the-device-and-user-verification-codes-from-github)

Reply:
Looking at the [rfc](https://datatracker.ietf.org/doc/html/rfc8628#section-3.2),

The seemed to be just using 32 random bytes instead of 16:

```elixir
iex(13)> :crypto.strong_rand_bytes(32) |> Base.url_encode64(padding: false)
"W_hqp7AGf3FfQzipLXQTC98Gur03ex8RE9h4a4oUYCM"
```

Reply:
For the user code, how about:

```elixir
:crypto.strong_rand_bytes(4) |> Base.hex_encode32(padding: false)
```

(I am using hex encoding to avoid `-` and `_`)

And for the hyphen in the middle

```elixir
:crypto.strong_rand_bytes(5) 
|> Base.hex_encode32(padding: false) 
|> String.split_at(4) 
|> then(fn {x, y} -> x <> "-" <> y end)
"QL0U-6B4M"
```